app-api/src/main/java/org/openmrs/eip/app/CustomDatabaseOffsetBackingStore.java

```diff
@@ -0,0 +1,165 @@
+			
+			if (log.isDebugEnabled()) {
+				log.debug("Loaded offset: {}", offset);
+			}
+			
+			Map<byte[], byte[]> raw = (Map<byte[], byte[]>) SerializationUtils.deserialize(offset.getData());
+			data = DebeziumOffsetUtil.offsetRawToData(raw);
+			
+		} else {
+			log.info("No offset was found in the database");
+		}
+	}
+	
+	/**
+	 * @see {@link io.debezium.embedded.EmbeddedEngine#EmbeddedEngine()}
+	 */
+	@Override
+	public void configure(WorkerConfig config) {
+		Map<String, Object> originals = config.originals();
+		
+		this.engineName = (String) originals.get(EmbeddedEngine.ENGINE_NAME.name());
+		String valueConverterClassName = (String) originals.get(WorkerConfig.VALUE_CONVERTER_CLASS_CONFIG);
+		
+		Map<String, Object> valueConverterOriginalConfs = config
+		        .originalsWithPrefix(WorkerConfig.INTERNAL_VALUE_CONVERTER_CLASS_CONFIG + ".", true);
+		valueConverterOriginalConfs.put(JsonConverterConfig.SCHEMAS_ENABLE_CONFIG, false);
+		
+		this.valueConverter = Instantiator.getInstance(valueConverterClassName, () -> this.getClass().getClassLoader(),
+		    null);
+		this.valueConverter.configure(valueConverterOriginalConfs, false);
+	}
+	
+	public synchronized static void enable() {
+		disabled = false;
+		if (log.isDebugEnabled()) {
+			log.debug("Enabled saving of offsets");
+		}
+	}
+	
+	public synchronized static void disable() {
+		disabled = true;
+		if (log.isDebugEnabled()) {
+			log.debug("Disabled saving of offsets");
+		}
+	}
+	
+	public synchronized static boolean isDisabled() {
+		return disabled;
+	}
+	
+	public synchronized static boolean isPaused() {
+		return paused;
+	}
+	
+	public synchronized static void pause() {
+		paused = true;
+		if (log.isDebugEnabled()) {
+			log.debug("Pause saving of offsets");
+		}
+	}
+	
+	public synchronized static void unpause() {
+		paused = false;
+		if (log.isDebugEnabled()) {
+			log.debug("Removing pause on saving of offsets");
+		}
+	}
+	
+	@Override
+	protected void save() {
+		synchronized (CustomDatabaseOffsetBackingStore.class) {
+			if (disabled || paused) {
+				if (paused) {
+					if (log.isDebugEnabled()) {
+						log.debug("Skipping saving of offset because it is paused");
+					}
+				} else {
+					log.warn("Skipping saving of offset because an error was encountered while processing a source record");
+				}
+				
+				return;
+			}
+			
+			if (log.isDebugEnabled()) {
+				log.debug("Saving binlog offset");
+			}
+			
+			doSave();
+		}
+	}
+	
+	protected void doSave() {
+		if (log.isDebugEnabled()) {
+			log.debug("Saving offset to the database");
+		}
+		
+		DebeziumOffset offset = new DebeziumOffset();
+		offset.setDateCreated(new Date());
+		offset.setData(SerializationUtils.serialize((Serializable) DebeziumOffsetUtil.offsetDataToRaw(data)));
+		offset.setBinlogFileName(DebeziumOffsetUtil.getOffsetBinlogFilename(data, valueConverter, engineName));
+		offset.setEnabled(Boolean.TRUE);
+		
+		offset = debeziumOffsetRepository.save(offset);
```

Comment:
I was expecting that here is where you always first validate the offset before saving it, that you have exactly one row in the database table